streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java

```diff
@@ -0,0 +1,411 @@
+
+    private ColumnFamilyHandle noTimestampColumnFamily;
+    private ColumnFamilyHandle withTimestampColumnFamily;
+
+    RocksDBTimestampedStore(final String name) {
+        super(name);
+    }
+
+    RocksDBTimestampedStore(final String name,
+                            final String parentDir) {
+        super(name, parentDir);
+    }
+
+    @Override
+    void openRocksDB(final DBOptions dbOptions,
+                     final ColumnFamilyOptions columnFamilyOptions) {
+        final List<ColumnFamilyDescriptor> columnFamilyDescriptors = asList(
+            new ColumnFamilyDescriptor(RocksDB.DEFAULT_COLUMN_FAMILY, columnFamilyOptions),
+            new ColumnFamilyDescriptor("keyValueWithTimestamp".getBytes(StandardCharsets.UTF_8), columnFamilyOptions));
+        final List<ColumnFamilyHandle> columnFamilies = new ArrayList<>(columnFamilyDescriptors.size());
+
+        try {
+            db = RocksDB.open(dbOptions, dbDir.getAbsolutePath(), columnFamilyDescriptors, columnFamilies);
+
+            noTimestampColumnFamily = columnFamilies.get(0);
+            withTimestampColumnFamily = columnFamilies.get(1);
+
+            final RocksIterator noTimestampsIter = db.newIterator(noTimestampColumnFamily);
+            noTimestampsIter.seekToFirst();
+            if (noTimestampsIter.isValid()) {
+                log.info("Opening store {} in upgrade mode", name);
+                dbAccessor = new DualColumnFamilyAccessor();
+            } else {
+                log.info("Opening store {} in regular mode", name);
+                dbAccessor = new WithTimestampColumnFamilyAccessor();
+            }
+            noTimestampsIter.close();
+        } catch (final RocksDBException e) {
+            if ("Column family not found: : keyValueWithTimestamp".equals(e.getMessage())) {
+                try {
+                    db = RocksDB.open(dbOptions, dbDir.getAbsolutePath(), columnFamilyDescriptors.subList(0, 1), columnFamilies);
+                    columnFamilies.add(db.createColumnFamily(columnFamilyDescriptors.get(1)));
+                } catch (final RocksDBException fatal) {
+                    throw new ProcessorStateException("Error opening store " + name + " at location " + dbDir.toString(), fatal);
+                }
+                noTimestampColumnFamily = columnFamilies.get(0);
+                withTimestampColumnFamily = columnFamilies.get(1);
+                log.info("Opening store {} in upgrade mode", name);
+                dbAccessor = new DualColumnFamilyAccessor();
+            } else {
+                throw new ProcessorStateException("Error opening store " + name + " at location " + dbDir.toString(), e);
+            }
+        }
+    }
+
+
+
+    private class WithTimestampColumnFamilyAccessor extends SingleColumnFamilyAccessor {
+        private WithTimestampColumnFamilyAccessor() {
+            super(withTimestampColumnFamily);
+        }
+    }
+
+    private class DualColumnFamilyAccessor implements RocksDBAccessor {
```

Comment:
If we make one, how does it now that it need to upgrade stuff or not? Seems to imply a `null` check for each operation if both or only one CF should be accessed? That would imply runtime overhead. Also, I _think_ the code would be a little bit harder to read. Thoughts?

Reply:
Yes.

Reply:
So that means we do not need to do extra check inside this accessor impl right?

Reply:
Personally, I like the existing design. The `DualColumnFamilyAccessor` has to do a lot of extra checking that isn't necessary if there's just one cf to deal with. If we collapse them into one class with a flag, we pay for it with a lot of branching.

One thing I did find confusing was reasoning about the fact that the Dual accessor is embedded in this (Timestamped) class, and the Single accessor is embedded in the parent (non-timestamped) class. But, we're using it as an accessor for this (the child) class. This seems unnecessarily convoluted, and it's a little hard to see if it's actually ok, or just coincidentally ok, since the parent and child APIs are only semantically, rather than actually, different.

It seems simpler to understand if we pull both accessors out into separate classes that take `db`, `name`, `options`, etc as constructor arguments, rather than closing over protected state.

Reply:
I looked at it closer. I still think it's better to split them out, but I also don't think it's a correctness issue right now, so I'd be fine with merging what you have.

Reply:
Let's to some refactoring as follow ups. It's internal anyway. Would like to keep the PRs focused to get stuff in and merged :)

Reply:
Actually I was only suggesting to make the current Dual accessor to be more general
currently it assumes the `old` to be `default`, and `new` to be `withTimestamp`.
what I was suggesting is only to make these two parameterized; so that in the future we only have two accessor impls:

1) XX-CF only; which we already do in this PR.
2) XX-to-YY upgrade: old XX CF to YY CF upgrade accessor.

All that being said, I'm okay with such refactoring as follow-ups.

Reply:
This refactoring is already done: `DualAccessor` has a constructor with two CF parameters.

There is one missing piece: the conversion function is hard coded. After `TimestampedByteStore` PR is merged, I will refactor this to pass in a `RecordConverter` that and will pass in `TimestampedByteStore#convertValueToNewFormat()`